### PR TITLE
Content storage

### PR DESCRIPTION
**Pull request by fearlessfe**
impl content storage #9

### CODE REVIEW

File: p2p/discover/portal_protocol.go [hunk 1]
```diff
@@ -62,6 +64,16 @@ const (
 	PersistOfferRequestKind   byte = 0x02
 )
 
+var ErrNilContentKey = errors.New("content key cannot be nil")
```
Comment (GrapeBaBa) [inline p2p/discover/portal_protocol.go:67]: shouldn't these all be moved to content_storage.go?

File: portalnetwork/storage/content_storage.go [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+}
+
+func (p *ContentStorage) batchDel(ids [][]byte) error {
+	query := "DELETE FROM kvstore WHERE key IN (?" + strings.Repeat(", ?", len(ids)-1) + ")"
+	args := make([]interface{}, len(ids))
+	for i, id := range ids {
+		args[i] = id
+	}
+
+	// 执行删除操作
```
Comment (GrapeBaBa) [inline portalnetwork/storage/content_storage.go:356]: don't use Chinese comment

File: portalnetwork/storage/content_storage_test.go [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+
+func genBytes(length int) []byte {
+	res := make([]byte, length)
+	for i := 0; i < length; i++ {
+		res[i] = byte(i)
+	}
+	return res
+}
+
+func defaultContentIdFunc(contentKey []byte) []byte {
```
Comment (GrapeBaBa) [inline portalnetwork/storage/content_storage_test.go:29]: Duplicate?

File: portalnetwork/storage/content_storage.go [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+	sqlite3 "github.com/mattn/go-sqlite3"
+)
+
+var (
+	ErrContentNotFound = fmt.Errorf("content not found")
+
+	maxDistance = uint256.MustFromHex("0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff")
+)
+
+const (
```
Comment (GrapeBaBa) [inline portalnetwork/storage/content_storage.go:24]: const before var

File: portalnetwork/storage/content_storage.go [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+	containSql                 = "SELECT 1 FROM kvstore WHERE key = (?1);"
+	getAllOrderedByDistanceSql = "SELECT key, length(value), xor(key, (?1)) as distance FROM kvstore ORDER BY distance DESC;"
+	deleteOutOfRadiusStmt      = "DELETE FROM kvstore WHERE greater(xor(key, (?1)), (?2)) = 1"
+	XOR_FIND_FARTHEST_QUERY    = `SELECT
+		xor(key, (?1)) as distance
+		FROM kvstore
+		ORDER BY distance DESC`
+)
+
+type ContentStorage struct {
```
Comment (GrapeBaBa) [inline portalnetwork/storage/content_storage.go:44]: public method comments not follow go standard

File: portalnetwork/storage/content_storage_test.go (hunk 3 above)
Comment (fearlessfe) [reply portalnetwork/storage/content_storage_test.go:29]: yes, I will delete it, the content storage not care about how the `contentKey` tranform to `contentId`

File: p2p/discover/portal_protocol.go [hunk 6]
```diff
@@ -130,12 +144,18 @@ type PortalProtocol struct {
...
 	cancelCloseCtx context.CancelFunc
-	storage        Storage
+	storage        ContentStorage
+	toContentId    func(contentKey []byte) []byte
 
 	contentQueue chan *ContentElement
 }
 
-func NewPortalProtocol(config *PortalProtocolConfig, protocolId string, privateKey *ecdsa.PrivateKey, storage Storage, contentQueue chan *ContentElement) (*PortalProtocol, error) {
+func defaultContentIdFunc(contentKey []byte) []byte {
```
[the remark this answers is not in the document]
Comment (fearlessfe) [reply p2p/discover/portal_protocol.go:153]: this should be in `PortalProtocol`, storage do not care about the `contentKey` to `contentId`

File: p2p/discover/portal_protocol.go (hunk 1 above)
Comment (fearlessfe) [reply p2p/discover/portal_protocol.go:67]: this err is return from the `PortalProtocol` before call the `content_storage`, so this should be in `PortalProtocol`

**Pull request merged**